crates/y-sweet-core/src/auth.rs

```diff
@@ -165,24 +129,81 @@ fn hash(bytes: &[u8]) -> Vec<u8> {
 impl Authenticator {
     pub fn new(private_key: &str) -> Result<Self, AuthError> {
         let private_key = b64_decode(private_key)?;
-        let server_token = Payload::new(Permission::Server).sign(&private_key);
 
         Ok(Self {
             private_key,
-            server_token,
+            key_id: None,
         })
     }
 
-    pub fn server_token(&self) -> &str {
-        &self.server_token
+    pub fn server_token(&self) -> String {
+        self.sign(Payload::new(Permission::Server))
+    }
+
+    fn sign(&self, payload: Payload) -> String {
+        let mut hash_payload =
+            bincode_encode(&payload).expect("Bincode serialization should not fail.");
+        hash_payload.extend_from_slice(&self.private_key);
+
+        let token = hash(&hash_payload);
+
+        let auth_req = AuthenticatedRequest { payload, token };
+
+        let auth_enc = bincode_encode(&auth_req).expect("Bincode serialization should not fail.");
+        let result = b64_encode(&auth_enc);
+        if let Some(key_id) = &self.key_id {
+            format!("{}.{}", key_id, result)
+        } else {
+            result
+        }
+    }
+
+    fn verify(&self, token: &str, current_time: u64) -> Result<Payload, AuthError> {
+        let token = if let Some((prefix, token)) = token.split_once('.') {
+            if Some(prefix) != self.key_id.as_deref() {
+                return Err(AuthError::KeyMismatch);
+            }
+
+            token
+        } else {
+            if self.key_id.is_some() {
+                return Err(AuthError::KeyMismatch);
+            }
+
+            &token
+        };
+
+        let auth_req: AuthenticatedRequest =
+            bincode_decode(&b64_decode(token)?).map_err(|_| AuthError::InvalidToken)?;
+
+        let mut payload =
+            bincode_encode(&auth_req.payload).expect("Bincode serialization should not fail.");
+        payload.extend_from_slice(&self.private_key);
+
+        let expected_token = hash(&payload);
+
+        if expected_token != auth_req.token {
+            Err(AuthError::InvalidSignature)
+        } else if auth_req.payload.expiration_millis.unwrap_or(u64::MAX) < current_time {
+            Err(AuthError::Expired)
+        } else {
+            Ok(auth_req.payload)
+        }
+    }
+
+    pub fn with_key_id(self, key_id: String) -> Self {
```

Comment:
does the key id itself not have to be url safe base 64?

Reply:
ah, it doesn't since it's in the header.

are there any limitations there? or can that be any utf-8

Reply:
It has a restricted alphabet, I guess it doesn’t hurt to check for that. Maybe I’ll restrict to url-safe base64